src/compiler/resolutionCache.ts

```diff
@@ -0,0 +1,316 @@
+/// <reference path="types.ts"/>
+/// <reference path="core.ts"/>
+
+namespace ts {
+    export interface ResolutionCache {
+        setModuleResolutionHost(host: ModuleResolutionHost): void;
+
+        startRecordingFilesWithChangedResolutions(): void;
+        finishRecordingFilesWithChangedResolutions(): Path[];
+
+        resolveModuleNames(moduleNames: string[], containingFile: string, logChanges: boolean): ResolvedModuleFull[];
+        resolveTypeReferenceDirectives(typeDirectiveNames: string[], containingFile: string): ResolvedTypeReferenceDirective[];
+
+        invalidateResolutionOfDeletedFile(filePath: Path): void;
+        invalidateResolutionOfChangedFailedLookupLocation(failedLookupLocation: string): void;
+
+        createHasInvalidatedResolution(): HasInvalidatedResolution;
+
+        clear(): void;
+    }
+
+    type NameResolutionWithFailedLookupLocations = { failedLookupLocations: string[], isInvalidated?: boolean };
+    type ResolverWithGlobalCache = (primaryResult: ResolvedModuleWithFailedLookupLocations, moduleName: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost) => ResolvedModuleWithFailedLookupLocations | undefined;
+
+    /*@internal*/
+    export function resolveWithGlobalCache(primaryResult: ResolvedModuleWithFailedLookupLocations, moduleName: string, compilerOptions: CompilerOptions, host: ModuleResolutionHost, globalCache: string | undefined, projectName: string): ResolvedModuleWithFailedLookupLocations | undefined {
+        if (!isExternalModuleNameRelative(moduleName) && !(primaryResult.resolvedModule && extensionIsTypeScript(primaryResult.resolvedModule.extension)) && globalCache !== undefined) {
+            // otherwise try to load typings from @types
+
+            // create different collection of failed lookup locations for second pass
+            // if it will fail and we've already found something during the first pass - we don't want to pollute its results
+            const { resolvedModule, failedLookupLocations } = loadModuleFromGlobalCache(moduleName, projectName, compilerOptions, host, globalCache);
+            if (resolvedModule) {
+                return { resolvedModule, failedLookupLocations: primaryResult.failedLookupLocations.concat(failedLookupLocations) };
+            }
+        }
+    }
+
+    /*@internal*/
+    export function createResolutionCache(
+        toPath: (fileName: string) => Path,
+        getCompilerOptions: () => CompilerOptions,
+        watchForFailedLookupLocation: (failedLookupLocation: string, containingFile: string, name: string) => FileWatcher,
+        log: (s: string) => void,
+        resolveWithGlobalCache?: ResolverWithGlobalCache): ResolutionCache {

```

Comment:
It looks like the `resolveWithGlobalCache` parmeter is always either undefined or the outer (shadowed) `resolveWithGlobalCache` function -- you could just take the extra parameters needed by that that instead, and call it directly.